### PR TITLE
Add a Jacobian-Free Newton-Krylov solver and preconditioning hooks

### PR DESCRIPTION
**Pull request by thomasgibson**
# Description
Co-authored by: @Zhengyu-Huang 

This PR adds a Jacobian-Free Newton-Krylov solver and adds hooks for using preconditioners. This is still a work in progress, but comments will be really appreciated.

Some things left to do:

- Clean up documentation and expand
- Clean up PC hooks (PC abstraction)

I have

- [ ] Written and run all necessary tests with CLIMA by including `tests/runtests.jl`
- [ ] Followed all necessary [style guidelines](https://clima.github.io/ClimateMachine.jl/latest/DevDocs/CodeStyle/) and run `julia .dev/climaformat.jl .`
- [ ] Updated the documentation to reflect changes from this PR.

# For review by CLIMA developers

- [ ] There are no open pull requests for this already
- [ ] CLIMA developers with relevant expertise have been assigned to review this submission
- [ ] The code conforms to the [style guidelines](https://clima.github.io/ClimateMachine.jl/latest/DevDocs/CodeStyle/) and has consistent naming conventions. `julia .dev/format.jl` has been run in a separate commit.
- [ ] This code does what it is technically intended to do (all numerics make sense physically and/or computationally)

### CODE REVIEW

**Comment by Zhengyu-Huang** [comment]
Some potential improvements
1. nonlinear implicit explicit split only works for Additive Runge Kutta with NaiveVariant(), but not LowStorageVariant()
2. Preconditioner construction can be improved. PC is computed with finite difference method.  For each column of the preconditioner matrix P,  P_i = [F(Q_c + ϵe_i) - F(Q_c)]/ϵ, here e_i = [0, ..0,1,0..0], here F(Q_c + ϵe_i) - F(Q_c)] is sparse (nonzero in several elements). (Jeremy)
3. The number of degrees of freedom is about O(600)

File: src/Numerics/ODESolvers/BackwardEulerSolvers.jl [hunk 1]
```diff
@@ -128,9 +172,157 @@ function update_backward_Euler_solver!(lin::LinBESolver, Q, α)
...
+    LinBESolver
+
+Concrete implementation of an `AbstractBackwardEulerSolver` to use nonlinear
+solvers of type `NLS`. See helper type
+[`NonLinearBackwardEulerSolver`](@ref)
+```
+    Q = Qhat + α f_imp(Q, param, time)
+```
+"""
+
```
Comment (charleskawczynski) [inline src/Numerics/ODESolvers/BackwardEulerSolvers.jl:245]:
```suggestion
```

File: src/Numerics/ODESolvers/BackwardEulerSolvers.jl [hunk 2]
```diff
@@ -128,9 +172,157 @@ function update_backward_Euler_solver!(lin::LinBESolver, Q, α)
...
+Returns a concrete implementation of an `AbstractBackwardEulerSolver` that will
+solve for `Q` in nonlinear systems of the form of
+```
+    Q = Qhat + α f(Q, param, time)
+```
+Create an empty JacobianAction 
+Create an empty preconditioner if preconditioner_update_freq > 0
+"""
+
```
Comment (charleskawczynski) [inline src/Numerics/ODESolvers/BackwardEulerSolvers.jl:277]:
```suggestion
```
These newlines, between the doc string and the function/struct, will break the doc build.

File: src/Numerics/SystemSolvers/jacobian_free_newton_krylov_solver.jl [hunk 3]
```diff
@@ -0,0 +1,228 @@
...
+form on a vector `ΔQ` using the difference quotient:
+
+      ∂F(Q)      F(Q + e ΔQ) - F(Q)
+JΔQ = ---- ΔQ ≈ -------------------
+       ∂Q                e
+
+
+Compute  JΔQ with cached Q and F(Q), and the direction  dQ
+"""
+
```
Comment (charleskawczynski) [inline src/Numerics/SystemSolvers/jacobian_free_newton_krylov_solver.jl:62]:
```suggestion
```

File: src/Numerics/SystemSolvers/jacobian_free_newton_krylov_solver.jl [hunk 4]
```diff
@@ -0,0 +1,228 @@
...
+
+rhs!           : nonlinear operator F(Q)
+ϵ::FT          : ϵ used for finite difference, e = e(Q, ϵ)
+Q::AT          : cache for Q
+Qdq::AT        : container for Q + ϵΔQ
+Fq::AT         : cache for F(Q) 
+Fqdq::AT       : container for F(Q + ϵΔQ)
+"""
+
```
Comment (charleskawczynski) [inline src/Numerics/SystemSolvers/jacobian_free_newton_krylov_solver.jl:29]:
```suggestion
```

File: src/Numerics/SystemSolvers/jacobian_free_newton_krylov_solver.jl [hunk 5]
```diff
@@ -0,0 +1,228 @@
...
+rhs!:  functor rhs!(Q) =  F(Q)
+jvp!:  Jacobian action jvp!(ΔQ)  = dF/dQ(Q) ⋅ ΔQ
+preconditioner: approximation of dF/dQ(Q)
+
+Q : Q^n
+Qrhs : Frhs
+solver: linear solver
+"""
+
```
Comment (charleskawczynski) [inline src/Numerics/SystemSolvers/jacobian_free_newton_krylov_solver.jl:190]:
```suggestion
```

File: src/Numerics/SystemSolvers/preconditioner.jl [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+
+
+
+"""
+mutable struct NoPreconditioner
+end
+
+Do nothing
+"""
+
```
Comment (charleskawczynski) [inline src/Numerics/SystemSolvers/preconditioner.jl:18]:
```suggestion
```

File: src/Numerics/ODESolvers/BackwardEulerSolvers.jl [hunk 7]
```diff
@@ -128,9 +172,147 @@ function update_backward_Euler_solver!(lin::LinBESolver, Q, α)
...
+isadjustable: TODO not used, might use for updating preconditioner
+preconditioner_update_freq:  relavent to Jacobian free -1: no preconditioner;
+                             positive number, update every freq times
+"""
+struct NonLinearBackwardEulerSolver{NLS}
+    nlsolver::NLS
+    isadjustable::Bool
+    # preconditioner_update_freq, -1: no preconditioner;
+    # positive number, update every freq times
+    preconditioner_update_freq::Int64
```
Comment (charleskawczynski) [inline src/Numerics/ODESolvers/BackwardEulerSolvers.jl:218]:
```suggestion
    preconditioner_update_freq::Int
```

File: src/Numerics/ODESolvers/BackwardEulerSolvers.jl [hunk 8]
```diff
@@ -128,9 +172,147 @@ function update_backward_Euler_solver!(lin::LinBESolver, Q, α)
...
+        linearsolve!(rhs!, lin.factors, lin.solver, Q, Qhat, p, t)
+    end
+end
+
+"""
+struct NonLinearBackwardEulerSolver{NLS}
+    nlsolver::NLS
+    isadjustable::Bool
+    preconditioner_update_freq::Int64
+end
```
Comment (charleskawczynski) [inline src/Numerics/ODESolvers/BackwardEulerSolvers.jl:203]:
```suggestion
    struct NonLinearBackwardEulerSolver{NLS}
        nlsolver::NLS
        isadjustable::Bool
        preconditioner_update_freq::Int64
    end
```
Documenter treats indented lines (with 4 spaces) specially.

File: src/Numerics/SystemSolvers/SystemSolvers.jl [hunk 9]
```diff
@@ -37,6 +40,129 @@ This is an abstract type representing a generic linear solver.
...
+        # ||Delta Q|| / ||Q|| ?
+        relresidual = residual_norm / initial_residual_norm
+        if relresidual < tol || residual_norm < tol
+            @info "Newton converged in $iters iterations!"
+            converged = true
+        end
+    end
+
+    converged ||
+    @warn "Nonlinear solver did not attain convergence after $iters iterations"
```
Comment (charleskawczynski) [inline src/Numerics/SystemSolvers/SystemSolvers.jl:160]: Is it okay to leave the line to execute on the following line like this?

File: src/Numerics/SystemSolvers/jacobian_free_newton_krylov_solver.jl [hunk 10]
```diff
@@ -0,0 +1,216 @@
...
+        ϵ,
+        similar(Q),
+        similar(Q),
+        similar(Q),
+        similar(Q),
+    )
+end
+
+"""
+Approximations the action of the Jacobian of a nonlinear
```
Comment (charleskawczynski) [inline src/Numerics/SystemSolvers/jacobian_free_newton_krylov_solver.jl:49]:
```suggestion
Approximates the action of the Jacobian of a nonlinear
```

File: src/Numerics/ODESolvers/BackwardEulerSolvers.jl [hunk 11]
```diff
@@ -88,8 +91,12 @@ time step size.
 struct LinearBackwardEulerSolver{LS}
     solver::LS
     isadjustable::Bool
-    LinearBackwardEulerSolver(solver; isadjustable = false) =
-        new{typeof(solver)}(solver, isadjustable)
+    preconditioner_update_freq::Int64
```
Comment (charleskawczynski) [inline src/Numerics/ODESolvers/BackwardEulerSolvers.jl:94]:
```suggestion
    preconditioner_update_freq::Int
```

File: src/Numerics/SystemSolvers/preconditioners.jl [hunk 12]
```diff
@@ -0,0 +1,134 @@
...
+counter: count the number of Newton, when counter > update_freq or counter < 0, update precondition
+update_freq: preconditioner update frequency
+
+"""
+mutable struct ColumnwiseLUPreconditioner{AT} <: AbstractPreconditioner
+    A::DGColumnBandedMatrix
+    Q::AT
+    PQ::AT
+    counter::Int64
+    update_freq::Int64
```
Comment (charleskawczynski) [inline src/Numerics/SystemSolvers/preconditioners.jl:62]:
```suggestion
    counter::Int
    update_freq::Int
```

File: src/Numerics/SystemSolvers/columnwise_lu_solver.jl [hunk 13]
```diff
@@ -311,8 +336,53 @@ function banded_matrix(
...
+    nrealelem = length(topology.realelems)
+    nvertelem = topology.stacksize
+    nhorzelem = div(nrealelem, nvertelem)
+
+    dim = dimensionality(grid)
+
+    Nqj = dim == 2 ? 1 : Nq
+
+    # TODO
+    # fill!(A, zero(FT))
```
Comment (charleskawczynski) [inline src/Numerics/SystemSolvers/columnwise_lu_solver.jl:379]: Should this be addressed?

File: src/Numerics/SystemSolvers/preconditioners.jl [hunk 14]
```diff
@@ -0,0 +1,134 @@
...
+build an empty ColumnwiseLUPreconditioner
+
+dg: DG model, use only the grid information
+Q0: MPIArray, use only its size
+counter = -1, which indicates the preconditioner is empty
+update_freq: preconditioner update frequency
+
+"""
+function ColumnwiseLUPreconditioner(dg, Q0, update_freq = 100)
+    single_column = false
```
Comment (charleskawczynski) [inline src/Numerics/SystemSolvers/preconditioners.jl:76]: Should we expose this? Or does that not make sense for `ColumnwiseLUPreconditioner`?

Comment (thomasgibson) [reply src/Numerics/SystemSolvers/preconditioners.jl:76]: This is one of the things I wanted to do. I think it should be exposed. But apparently people want this in fast... So I will wait to do this in a later PR.

File: src/Numerics/ODESolvers/BackwardEulerSolvers.jl [hunk 15]
```diff
@@ -128,9 +172,147 @@ function update_backward_Euler_solver!(lin::LinBESolver, Q, α)
...
+Δt_is_adjustable(nlsolver::NonLinBESolver) = nlsolver.isadjustable
+
+"""
+    setup_backward_Euler_solver(solver::NonLinearBackwardEulerSolver, Q, α, tendency!)
+
+Returns a concrete implementation of an `AbstractBackwardEulerSolver` that will
+solve for `Q` in nonlinear systems of the form of
+```
+    Q = Qhat + α f(Q, param, time)
+```
```
Comment (simonbyrne) [inline src/Numerics/ODESolvers/BackwardEulerSolvers.jl:265]: backticks are redundant with indentantion
```suggestion

    Q = Qhat + α f(Q, param, time)

```

**Review by simonbyrne — APPROVED** [review]
Docs still need a bit of polish, but we can do that in the doc sprint

**Comment by jakebolewski** [comment]
@Zhengyu-Huang can you try again?

**Pull request merged**